docs/grammar/Solidity.g4

```diff
@@ -0,0 +1,489 @@
+ * The arguments are either given as comma separated list or as map of named arguments.
+ */
+callArgumentList: LParen ((expression (Comma expression)*)? | LBrace (namedArgument (Comma namedArgument)*)? RBrace) RParen;
+/**
+ * Qualified name of a user defined type.
+ */
+userDefinedTypeName: identifier (Period identifier)*;
+
+/**
+ * Call to a modifier. If the modifier takes no arguments, the argument list can be skipped entirely
+ * (including opening and closing parentheses).
+ */
+modifierInvocation: identifier callArgumentList?;
+/**
+ * Visibility for functions and function types.
+ */
+visibility: Internal | External | Private | Public;
+/**
+ * A list of parameters, such as function arguments or return values.
+ */
+parameterList: parameters+=parameterDeclaration (Comma parameters+=parameterDeclaration)*;
+//@doc:inline
+parameterDeclaration: type=typeName location=storageLocation? name=identifier?;
+/**
+ * Definition of a constructor.
+ * Must always supply an implementation.
+ * Note that specifying internal or public visibility is deprecated.
+ */
+constructorDefinition:
+    Constructor LParen (arguments=parameterList)? RParen
+    (modifierInvocation | Payable | Internal | Public)*
+    body=block;
+
+/**
+ * State mutability for function types.
+ * The default mutability 'non-payable' is assumed if no mutability is specified.
+ */
+stateMutability: Pure | View | Payable;
+/**
+ * An override specifier used for functions, modifiers or state variables.
+ * In cases where there are ambiguous declarations in several base contracts being overridden,
+ * a complete list of base contracts has to be given.
+ */
+overrideSpecifier: Override (LParen overrides+=userDefinedTypeName (Comma overrides+=userDefinedTypeName)* RParen)?;
+/**
+ * The definition of contract, library and interface functions.
+ * Depending on the context in which the function is defined, further restrictions may apply,
+ * e.g. functions in interfaces have to be unimplemented, i.e. may not contain a body block.
+ */
+functionDefinition
+locals[
+    boolean visibilitySet = false,
+    boolean mutabilitySet = false,
+    boolean virtualSet = false,
+    boolean overrideSpecifierSet = false
+]
+:
+	Function (identifier | Fallback | Receive)
+	LParen (arguments=parameterList)? RParen
+	(
+	    {!$visibilitySet}? visibility {$visibilitySet = true;}
+	    | {!$mutabilitySet}? stateMutability {$mutabilitySet = true;}
+	    | modifierInvocation
+	    | {!$virtualSet}? Virtual {$virtualSet = true;}
+	    | {!$overrideSpecifierSet}? overrideSpecifier {$overrideSpecifierSet = true;}
+	 )*
+	(Returns LParen returnParameters=parameterList RParen)?
+	(Semicolon | body=block);
+/**
+ * The definition of a modifier.
+ * Note that within the body block of a modifier, the underscore cannot be used as identifier,
+ * but is used as placeholder statement for the body of a function to which the modifier is applied.
+ */
+modifierDefinition
```

Comment:
It doesn't really enforce `_` to be present.

Reply:
So far that's an analysis thing (it's a SyntaxError, not a ParserError).